### PR TITLE
Fix #101, fix #102: core data unit tests.

### LINKED ISSUE #101: Unit Tests: CoreData
**CoreDataStackTests**

- [ ] testWorkerToMainMerge
- [ ] testMainToWorkerMerge
- [ ] testSaveContext

**CD Utilities:**

- [ ] compareContexts // will validate that both new workerContext, and mainContext have same data

### LINKED ISSUE #102: Unit Tests: Bookmarks
## Files:

**BookmarkTests**

- [ ] testAddFolder
- [ ] testAddBookmark
- [ ] testAddNestedBookmark
- [ ] testReordering
- [ ] testRemoveBookmark
- [ ] testRemoveFolder
- [ ] testRemoveFolderWithBookmarks
- [ ] testEditBookmark
- [ ] testEditFolder
- [ ] testEditTopLevelFolder

### PR DESCRIPTION
**Pull request by iccub**
- [x] `DataController` tests
- [x] `Bookmark` tests
- [x] `TabMO` tests
- [x] `Device` tests
- [x] `History` tests
- [x] `Domain` tests
- [x] `FaviconMO` tests

## Pull Request Checklist

- [x] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [x] I have updated the *Unit Tests* to cover new or changed functionality
- [x] I have updated the *UI Tests* to cover new or changed functionality
- [x] I have marked the bug with `[needsuplift]`
- [x] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

File: Data/models/DataController.swift [hunk 1]
```diff
@@ -30,7 +30,7 @@ private let log = Logger.browserLogger
 // Follow the stack design from http://floriankugler.com/2013/04/02/the-concurrent-core-data-stack/
 
 public class DataController: NSObject {
-    public static let shared = DataController()
+    public static var shared: DataController? = DataController()
```
Comment (jhreis) [inline Data/models/DataController.swift]: What's the reason for making this an optional? Even if we allow it to be reset, we don't need to make it an optional. Just not a constant.

File: Data/models/DataController.swift [hunk 2]
```diff
@@ -146,14 +153,24 @@ public class DataController: NSObject {
             }
         }
     }
-}
-
-extension NSManagedObjectContext {
-    static var mainThreadContext: NSManagedObjectContext {
-        return DataController.shared.mainThreadContext
+    
+    public static func resetDatabase() {
```
Comment (jhreis) [inline Data/models/DataController.swift]: This should probably be in a more unit test environment, putting unit test functions in general classes can get pretty messy.

File: Data/models/Device.swift [hunk 3]
```diff
@@ -110,4 +110,13 @@ class Device: NSManagedObject, Syncable {
         }
     }
     
+    // Needed for testing only. No need to set it public, @testable annotation allows use of internal methods.
+    class func clearSharedDevice() {
```
Comment (jhreis) [inline Data/models/Device.swift]: Same, why does this need to be here instead of in the unit test target?

**Review by jhreis — CHANGES_REQUESTED** [review]
Coverage lgtm, just some more general stuff in the normal codebase.

File: Data/models/DataController.swift (hunk 2 above)
Comment (iccub) [reply Data/models/DataController.swift]: I would need to make initializer at least `internal` then. If you prefer it this way I can change it.

File: Data/models/Device.swift (hunk 3 above)
Comment (iccub) [reply Data/models/Device.swift]: Same goes here, I need to drop `private` access for `shared`

**Review by jhreis — APPROVED** [review]

**Pull request merged**